### PR TITLE
[macOS] -_windowSnapshotInRect:withOptions: should be marked NS_RETURNS_RETAINED

### PR DESCRIPTION
**Pull request by whsieh**
#### 590c7af23a413f38d58531310cdd19dc83e72715
<pre>
[macOS] -_windowSnapshotInRect:withOptions: should be marked NS_RETURNS_RETAINED
<a href="https://bugs.webkit.org/show_bug.cgi?id=282667">https://bugs.webkit.org/show_bug.cgi?id=282667</a>

Reviewed by Abrar Rahman Protyasha and Aditya Keerthi.

* Source/WebCore/platform/graphics/cg/CGWindowUtilities.cpp:
(WebCore::cgWindowListCreateImage):

Drive-by fix: make this return a `RetainPtr`, to clarify ownership semantics when calling into this
function.

* Source/WebCore/platform/graphics/cg/CGWindowUtilities.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:

Add `NS_RETURNS_RETAINED` to clarify that this SPI method returns a +1 object.

* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.mm:
(-[WKFullScreenWindowController enterFullScreen:]):
(takeWindowSnapshot):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::takeWindowSnapshot):
* Source/WebKitLegacy/mac/WebView/WebFullScreenController.mm:
(-[WebFullScreenController enterFullScreen:]):
* Source/WebKitLegacy/mac/WebView/WebViewPrivate.h:
* Tools/DumpRenderTree/mac/PixelDumpSupportMac.mm:
(takeWindowSnapshot):
* Tools/WebKitTestRunner/mac/PlatformWebViewMac.mm:
(WTR::PlatformWebView::windowSnapshotImage):

Fix a bad call to `adoptCF` around an unretained pointer.

Canonical link: <a href="https://commits.webkit.org/286226@main">https://commits.webkit.org/286226@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/be7ca2acd0e2c603067543934d0b4b11a95f31c6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75246 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54685 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28086 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79707 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26502 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77363 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63825 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2470 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59056 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17297 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78313 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49233 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64641 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39431 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46622 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22137 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24830 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/67662 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22476 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81192 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2576 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1596 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67305 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2726 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64640 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66596 "Found 2 new API test failures: /TestWebKit:WebKit.GetInjectedBundleInitializationUserDataCallback, /TestWebKit:WebKit.InjectedBundleInitializationUserDataCallbackWins (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16565 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10544 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8703 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2537 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5342 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2562 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3491 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2571 "Built successfully") | | |

### CODE REVIEW

**Review by aprotyas — APPROVED** [review]

**Review by pxlcoder — APPROVED** [review]

**Comment by whsieh** [comment]
Interesting... the test failures actually seem pertinent:

```
Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   CoreFoundation                	       0x18668bd60 _CFRelease.cold.3 + 84
1   CoreFoundation                	       0x1865fa944 _CFRelease + 488
2   AppKit                        	       0x189872e30 -[NSCGImageSnapshotRep dealloc] + 44
3   CoreFoundation                	       0x1864ead74 -[__NSSingleObjectArrayI dealloc] + 64
4   AppKit                        	       0x18a01daa0 -[NSImageArrayRepProvider dealloc] + 32
5   libobjc.A.dylib               	       0x1860b3bd4 AutoreleasePoolPage::releaseUntil(objc_object**) + 196
6   libobjc.A.dylib               	       0x1860b079c objc_autoreleasePoolPop + 256
7   JavaScriptCore                	       0x1064cd114 WTF::AutodrainedPool::~AutodrainedPool() + 24
8   WebKitTestRunner              	       0x100d52ba0 WTR::TestController::runTest(char const*) + 616 (TestController.cpp:1727)
9   WebKitTestRunner              	       0x100d52e38 WTR::TestController::runTestingServerLoop() + 316 (TestController.cpp:1769)
10  WebKitTestRunner              	       0x100d4bf5c WTR::TestController::run() + 16 (TestController.cpp:1777) [inlined]
11  WebKitTestRunner              	       0x100d4bf5c WTR::TestController::TestController(int, char const**) + 388 (TestController.cpp:213)
12  WebKitTestRunner              	       0x100d4c450 WTR::TestController::TestController(int, char const**) + 12 (TestController.cpp:210)
13  WebKitTestRunner              	       0x100d2b748 main + 140 (main.mm:77)
```

It looks like `CGWindowListCreateImage` might not actually return a +1 object, despite the name 🤔

**Comment by whsieh** [comment]
Yeah.. looks like `CGWindowListCreateImage` returns an autoreleased (not +1) object, so it should not be adopted.

I'll just use this bug to track adding `NS_RETURNS_RETAINED` then.

**Comment by whsieh** [comment]
> Yeah.. looks like `CGWindowListCreateImage` returns an autoreleased (not +1) object, so it should not be adopted.

After some more debugging, I realized that `CGWindowListCreateImage` _does_ return a +1 object after all — the real issue is that there's yet another memory management bug that was dependent on us leaking the CGImage in `WebCore::cgWindowListCreateImage`:

```
RetainPtr<CGImageRef> PlatformWebView::windowSnapshotImage()
{
…
    RetainPtr image = adoptNS([platformView() _windowSnapshotInRect:CGRectNull withOptions:options]);
    return adoptCF([image CGImageForProposedRect:nil context:nil hints:nil]);
}
```

The `adoptCF` at the end is incorrect, since `-CGImageForProposedRect:context:hints:` does not return a +1 object (it's actually annotated with `CF_RETURNS_NOT_RETAINED`).

So basically, the lack of adoption (leak) in `_windowSnapshotInRect:withOptions:` right now is being made up for by the extra adoption (overrelease) in `PlatformWebView::windowSnapshotImage`…

I'll fix this as well.
